tetraelf-hierarchy/hierarchy-tool/tool/src/java/org/sakaiproject/hierarchy/tool/entityproviders/HierarchyEntityProviderImpl.java

```diff
@@ -98,4 +98,47 @@ else if(nodeChild instanceof PortalNodeSite){
 		}
 		return hierarchySiteSummary;
 	}
-}
+
+	/**
+	 * /portal-hierarchy/locations.json?siteid=id)
+	 */
+	@EntityCustomAction(action = "locations", viewKey = EntityView.VIEW_LIST)
+	public List<PortalNodeSummary> getNodesFromRootForSite(EntityView view, Map<String, Object> params){
+		String siteId =  (String)params.get("siteid");
+
+		if(StringUtils.isBlank(siteId)){
+			throw new IllegalArgumentException("site ID must be set in order to get the site node locations");
+		}
+
+		List<PortalNodeSummary> portalNodeSummaries = new ArrayList<>();
+
+		if (portalHierarchyService.siteExists(siteId)) {
+
+			List<PortalNode> portalNodes = portalHierarchyService.getNodesWithSite(siteId);
+			PortalNode portalNodePrimary = portalHierarchyService.getDefaultNode(siteId);
+
+			// Site exists but if it is not in the hierarchy show an empty entity.
+			if (portalNodes.isEmpty()) {
+				return portalNodeSummaries;
+			}
+
+			for (PortalNode pn : portalNodes) {
+				if (pn.canView()) {
+					PortalNodeSummary pne = new PortalNodeSummary(pn);
+					if ((pn.getId()).equals(portalNodePrimary.getId())) {
+						pne.setPrimary(true);
+					}
+					portalNodeSummaries.add(pne);
+				}
+			}
+
+			if (portalNodeSummaries.size() == 0) {
+				//Site exists but user does not have permission to view any hierarchy sites that contain the requesting site. 
+				throw new IllegalArgumentException("You do not have permission to view any pages that include this site");
```

Comment:
It looks to me as if this check isn't necessary, if one can see a site one seems to be able to see it anywhere in the hierarchy, even if one doesn't have access to the parent site, is this what you think is the case Matthew?